src/vm/zapsig.cpp

```diff
@@ -1106,6 +1106,77 @@ FieldDesc * ZapSig::DecodeField(Module *pReferencingModule,
     return pField;
 }
 
+// Copy single type signature, adding ELEMENT_TYPE_MODULE_ZAPSIG to types that are encoded using tokens.
+// The source signature originates from the module with index specified by the parameter moduleIndex.
+// Passing moduleIndex set to MODULE_INDEX_NONE results in pure copy of the signature.
+//
+void ZapSig::CopyTypeSignature(SigParser* pSigParser, SigBuilder* pSigBuilder, DWORD moduleIndex)
+{
+    if (moduleIndex != MODULE_INDEX_NONE)
+    {
+        SigParser peekParser(*pSigParser);
+
+        BYTE byteType;
+        IfFailThrow(peekParser.GetByte(&byteType));
+        CorElementType type = (CorElementType)byteType;
+
+        // Some element types have sub-type that is what we are interested in
+        switch ((DWORD)type)
+        {
+            case ELEMENT_TYPE_BYREF:
+            case ELEMENT_TYPE_PTR:
+            case ELEMENT_TYPE_PINNED:
```

Comment:
My understanding is that this will be used only for generic type arguments and such. I wouldn't expect to see byref/ptr/pinned in any of those cases. Could you assert false in these cases?

Also, I see that what has been implemented is a conservative approach to the problem, where if the signature isn't recognized as safe, it always adds the ELEMENT_TYPE_MODULE_ZAPSIG. I believe this is ok, but I'd like a comment describing the conservative nature of this code..